### PR TITLE
WFLY-10172 Add support for JSR-375 EE Security

### PR DESCRIPTION
**Pull request by stuartwdouglas**
https://issues.jboss.org/browse/WFLY-10172

### CODE REVIEW

**Review by darranl — APPROVED** [review]

File: ee-security/src/main/java/org/wildfly/extension/eesecurity/EESecurityDependencyProcessor.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+class EESecurityDependencyProcessor implements DeploymentUnitProcessor {
+    @Override
+    public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitProcessingException {
+
+        final DeploymentUnit unit = phaseContext.getDeploymentUnit();
+        final DeploymentUnit top = unit.getParent() == null ? unit : unit.getParent();
+        final ModuleLoader moduleLoader = Module.getBootModuleLoader();
+
+        final ModuleSpecification moduleSpec = unit.getAttachment(Attachments.MODULE_SPECIFICATION);
+        moduleSpec.addSystemDependency(new ModuleDependency(moduleLoader, ModuleIdentifier.fromString("javax.security.enterprise.api"), false, false, true, false));
```
Comment (mchoma) [inline ee-security/src/main/java/org/wildfly/extension/eesecurity/EESecurityDependencyProcessor.java]: Why is this necessary always add this dependency? Can't be also in if (securityPresent) ?

File: weld/subsystem/src/main/java/org/jboss/as/weld/WeldSubsystemAdd.java [hunk 2]
```diff
@@ -67,6 +74,30 @@
 
     static final WeldSubsystemAdd INSTANCE = new WeldSubsystemAdd();
 
+    static {
```
Comment (mchoma) [inline weld/subsystem/src/main/java/org/jboss/as/weld/WeldSubsystemAdd.java]: Can you elaborate why this logic is necessary in conjuction with this change?

File: ee-security/src/main/java/org/wildfly/extension/eesecurity/EESecurityDependencyProcessor.java (hunk 1 above)
Comment (stuartwdouglas) [reply ee-security/src/main/java/org/wildfly/extension/eesecurity/EESecurityDependencyProcessor.java]: Its required by the EE spec.

File: weld/subsystem/src/main/java/org/jboss/as/weld/WeldSubsystemAdd.java (hunk 2 above)
Comment (stuartwdouglas) [reply weld/subsystem/src/main/java/org/jboss/as/weld/WeldSubsystemAdd.java]: This extension uses CDI.current() instead of java:comp:BeanManager. CDI.current() presently just returns the root BM of the EAR, which does not have visibility to beans defined in a WAR. 

This change means that the extension will find the appropriate beans, even in the presence of  multiple WARs.

Really this could probably be considered a seperate bug that could be fixed independently.

Comment (mchoma) [reply weld/subsystem/src/main/java/org/jboss/as/weld/WeldSubsystemAdd.java]: Probably it would be better to treat as separate WELD bug to have visibility/attention of weld experts.

**Review by dmlloyd — APPROVED** [review]

File: ee-security/src/main/java/org/wildfly/extension/eesecurity/EESecurityDependencyProcessor.java (hunk 1 above)
Comment (stuartwdouglas) [reply ee-security/src/main/java/org/wildfly/extension/eesecurity/EESecurityDependencyProcessor.java]: To elaborate on this a bit, according to the platform spec all the API's have to be available to an application, even if they are not using them.

Comment (mchoma) [reply ee-security/src/main/java/org/wildfly/extension/eesecurity/EESecurityDependencyProcessor.java]: Thanks for clarification

File: weld/subsystem/src/main/java/org/jboss/as/weld/WeldProvider.java [hunk 3]
```diff
@@ -87,9 +90,23 @@ public CdiImpl(Container container, BeanManagerProxy rootBeanManager, WeldDeploy
         }
 
         @Override
-        public BeanManagerProxy getBeanManager() {
+        public BeanManager getBeanManager() {
             checkContainerState(container);
             final String callerName = getCallingClassName();
+            if(callerName.startsWith("org.glassfish.soteria")) {
```
Comment (mchoma) [inline weld/subsystem/src/main/java/org/jboss/as/weld/WeldProvider.java]: Definitelly safer to limit this behaviour change only to soteria.

File: weld/subsystem/src/main/java/org/jboss/as/weld/WeldProvider.java [hunk 4]
```diff
@@ -87,9 +90,23 @@ public CdiImpl(Container container, BeanManagerProxy rootBeanManager, WeldDeploy
...
+                //the Java EE Security RI uses CDI.current() to perform bean lookup, however
```
Comment (mchoma) [inline weld/subsystem/src/main/java/org/jboss/as/weld/WeldProvider.java]: This whole seems to me as workaround. What would be proper fix for this? Make soteria not to use CDI.current()? Or is there some design problem in WildFly?

Do we have weld issue for this to make proper discussion?

Comment (stuartwdouglas) [reply weld/subsystem/src/main/java/org/jboss/as/weld/WeldProvider.java]: CDI.current() is not very well specified, it does not really say how been visibility sour be handled.
